diff -ruN a/util/log.c b/util/log.c
--- a/util/log.c	2020-09-10 09:24:42.170106496 +0000
+++ b/util/log.c	2020-09-10 09:25:33.946107957 +0000
@@ -118,7 +118,7 @@
 		 * chroot and no longer be able to access dev/log and so on */
 		/* the facility is LOG_DAEMON by default, but
 		 * --with-syslog-facility=LOCAL[0-7] can override it */
-		openlog(ident, LOG_NDELAY, UB_SYSLOG_FACILITY);
+		openlog(ident, LOG_NDELAY, LOG_LOCAL0);
 		logging_to_syslog = 1;
 		lock_basic_unlock(&log_lock);
 		return;
