Index: postfix-3.0.3/src/xsasl/xsasl_cyrus_server.c
===================================================================
--- postfix-3.0.3.orig/src/xsasl/xsasl_cyrus_server.c	2016-01-05 19:29:08.981936443 +0000
+++ postfix-3.0.3/src/xsasl/xsasl_cyrus_server.c	2016-01-05 23:05:06.273454944 +0000
@@ -291,23 +291,12 @@
 #define NO_SESSION_CALLBACKS	((sasl_callback_t *) 0)
 #define NO_AUTH_REALM		((char *) 0)
 
-#if SASL_VERSION_MAJOR >= 2 && defined(USE_SASL_IP_AUTH)
-
-    /*
-     * Get IP addresses of local and remote endpoints for SASL.
-     */
-#error "USE_SASL_IP_AUTH is not implemented"
-
-#else
-
-    /*
-     * Don't give any IP address information to SASL.  SASLv1 doesn't use it,
-     * and in SASLv2 this will disable any mechanisms that do.
-     */
-    server_address = 0;
-    client_address = 0;
-#endif
 
+    server_address = strlen(args->server_addr) ? args->server_addr : "127.0.0.1";
+    client_address = strlen(args->client_addr) ? args->client_addr : "127.0.0.1";
+    if (msg_verbose)
+    	msg_info("%s: args->server_addr=%s, args->client_addr=%s, server_address=%s, client_address=%s",
+         myname, args->server_addr, args->client_addr, server_address, client_address);
     if ((sasl_status =
 	 SASL_SERVER_NEW(args->service, var_myhostname,
 			 args->user_realm ? args->user_realm : NO_AUTH_REALM,
