--- openldap-2.4.46/servers/slapd/back-mdb/config.c.orig	2018-08-31 21:34:31.111309728 +0000
+++ openldap-2.4.46/servers/slapd/back-mdb/config.c	2018-08-31 21:22:48.499227430 +0000
@@ -423,10 +435,11 @@
 			if ( c->valx == -1 ) {
 				int i;
 
-				/* delete all (FIXME) */
+				/* delete all */
 				for ( i = 0; i < mdb->mi_nattrs; i++ ) {
 					mdb->mi_attrs[i]->ai_indexmask |= MDB_INDEX_DELETING;
 				}
+				mdb->mi_defaultmask = 0;
 				mdb->mi_flags |= MDB_DEL_INDEX;
 				c->cleanup = mdb_cf_cleanup;
 
