From eaee3b39ba58bf0630679a35e7be6062d8cbd78c Mon Sep 17 00:00:00 2001
From: Howard Chu <hyc@openldap.org>
Date: Fri, 15 Jan 2016 10:37:56 +0000
Subject: [PATCH] ITS#8351 fix accesslog callback init

---
 servers/slapd/overlays/accesslog.c |    8 +++-----
 1 file changed, 3 insertions(+), 5 deletions(-)

diff --git a/servers/slapd/overlays/accesslog.c b/servers/slapd/overlays/accesslog.c
index e708794..992d043 100644
--- a/servers/slapd/overlays/accesslog.c
+++ b/servers/slapd/overlays/accesslog.c
@@ -578,7 +578,7 @@ log_age_unparse( int age, struct berval *agebv, size_t size )
 	agebv->bv_len = ptr - agebv->bv_val;
 }
 
-static slap_callback nullsc = { NULL, NULL, NULL, NULL };
+static slap_callback nullsc;
 
 #define PURGE_INCREMENT	100
 
@@ -637,7 +637,7 @@ accesslog_purge( void *ctx, void *arg )
 	OperationBuffer opbuf;
 	Operation *op;
 	SlapReply rs = {REP_RESULT};
-	slap_callback cb = { NULL, log_old_lookup, NULL, NULL };
+	slap_callback cb = { NULL, log_old_lookup, NULL, NULL, NULL };
 	Filter f;
 	AttributeAssertion ava = ATTRIBUTEASSERTION_INIT;
 	purge_data pd = {0};
@@ -1975,11 +1975,9 @@ accesslog_op_mod( Operation *op, SlapReply *rs )
 	}
 			
 	if ( doit ) {
-		slap_callback *cb = op->o_tmpalloc( sizeof( slap_callback ), op->o_tmpmemctx ), *cb2;
+		slap_callback *cb = op->o_tmpcalloc( 1, sizeof( slap_callback ), op->o_tmpmemctx ), *cb2;
 		cb->sc_cleanup = accesslog_mod_cleanup;
-		cb->sc_response = NULL;
 		cb->sc_private = on;
-		cb->sc_next = NULL;
 		for ( cb2 = op->o_callback; cb2->sc_next; cb2 = cb2->sc_next );
 		cb2->sc_next = cb;
 
-- 
1.7.10.4

