From 0ba50a1d064ef71145fe0a08d813d710634a23b4 Mon Sep 17 00:00:00 2001
From: Howard Chu <hyc@openldap.org>
Date: Wed, 2 May 2018 16:51:49 +0100
Subject: [PATCH] ITS#8843 check for NULL modlist

---
 servers/slapd/overlays/syncprov.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/servers/slapd/overlays/syncprov.c b/servers/slapd/overlays/syncprov.c
index 48ca2ae..160b57a 100644
--- a/servers/slapd/overlays/syncprov.c
+++ b/servers/slapd/overlays/syncprov.c
@@ -1936,6 +1936,7 @@ syncprov_op_response( Operation *op, SlapReply *rs )
 		/* Don't do any processing for consumer contextCSN updates */
 		if ( SLAPD_SYNC_IS_SYNCCONN( op->o_connid ) &&
 			op->o_tag == LDAP_REQ_MODIFY &&
+			op->orm_modlist &&
 			op->orm_modlist->sml_op == LDAP_MOD_REPLACE &&
 			op->orm_modlist->sml_desc == slap_schema.si_ad_contextCSN ) {
 			/* Catch contextCSN updates from syncrepl. We have to look at
-- 
2.7.4

