diff -ruN a/bin/amavis-mc b/bin/amavis-mc
--- a/bin/amavis-mc	2023-03-10 13:45:07.198404133 +0000
+++ b/bin/amavis-mc	2023-03-10 13:47:17.553885205 +0000
@@ -48,6 +48,7 @@
 use warnings;
 use warnings FATAL => qw(utf8 void);
 no warnings 'uninitialized';
+use lib "/opt/zimbra/common/lib/perl5";
 
 use vars qw($VERSION);  $VERSION = 2.008002;
 
@@ -73,16 +74,16 @@
 
 ### USER CONFIGURABLE:
 
-$daemon_user   = 'vscan';
-@daemon_groups = 'vscan';
+$daemon_user   = 'zimbra';
+@daemon_groups = 'zimbra';
 
-$pid_file = '/var/amavis/amavis-mc.pid';
+$pid_file = '/opt/zimbra/log/amavis-mc.pid';
 
 $log_level = 0;
 $syslog_ident = 'amavis-mc';
 $syslog_facility = LOG_MAIL;
 
-@path = qw(/usr/local/sbin /usr/local/bin /usr/sbin /sbin /usr/bin /bin);
+@path = qw(/opt/zimbra/common/sbin /usr/local/sbin /usr/local/bin /usr/sbin /sbin /usr/bin /bin);
 
 @services = (
   { cmd => 'amavis-services msg-forwarder' },
