diff -ruN a/bin/amavisd-snmp-subagent-zmq b/bin/amavisd-snmp-subagent-zmq
--- a/bin/amavisd-snmp-subagent-zmq	2023-03-10 13:51:30.337037837 +0000
+++ b/bin/amavisd-snmp-subagent-zmq	2023-03-10 13:51:57.692958783 +0000
@@ -49,6 +49,7 @@
 use warnings;
 use warnings FATAL => qw(utf8 void);
 no warnings 'uninitialized';
+use lib "/opt/zimbra/common/lib/perl5";
 
 use Errno qw(ESRCH ENOENT EACCES EEXIST);
 use POSIX ();
diff -ruN a/bin/amavisd-status b/bin/amavisd-status
--- a/bin/amavisd-status	2023-03-10 13:50:40.969190598 +0000
+++ b/bin/amavisd-status	2023-03-10 13:51:13.905083999 +0000
@@ -48,6 +48,7 @@
 use warnings;
 use warnings FATAL => qw(utf8 void);
 no warnings 'uninitialized';
+use lib "/opt/zimbra/common/lib/perl5";
 
 use Errno qw(ESRCH ENOENT);
 use POSIX qw(strftime);
