diff -ruN a/amavisd b/amavisd
--- a/amavisd	2022-08-22 09:25:28.945660379 +0000
+++ b/amavisd	2022-08-22 09:26:59.919428184 +0000
@@ -28124,7 +28124,7 @@
         my $f = $bare_fnames->[$k];  my $multi = 0;
         if ($one_at_a_time) {  # glob templates may be substrings anywhere
           local($1);  @query_expanded = @query_template;  # start afresh
-          s{ ( {} (?: / \* )? | \* ) }
+	   s{ ( \{\} (?: / \* )? | \* ) }
            { $1 eq '{}'   ? "$tempdir/parts"
            : $1 eq '{}/*' ? ($multi=1,"$tempdir/parts/$f")
            : $1 eq '*'    ? ($multi=1,$f)  : $1
