diff -ruN a/bin/amavis-services b/bin/amavis-services
--- a/bin/amavis-services	2023-03-10 14:02:55.403699797 +0000
+++ b/bin/amavis-services	2023-03-10 14:04:47.971596373 +0000
@@ -49,6 +49,7 @@
 no warnings 'uninitialized';
 
 use vars qw($VERSION);  $VERSION = 2.011002;
+use lib "/opt/zimbra/common/lib/perl5";
 
 use Errno qw(ESRCH ENOENT);
 use POSIX qw(strftime);
@@ -76,7 +77,8 @@
 
 # A socket to which amavisd child processes report their data.
 # should match one of the sockets in @zmq_sockets in amavisd.conf
-$inner_sock_specs = "ipc://$MYHOME/amavisd-zmq.sock";
+#$inner_sock_specs = "ipc://$MYHOME/amavisd-zmq.sock";
+$inner_sock_specs = ( "ipc:///opt/zimbra/data/tmp/amavisd-zmq.sock" );
 
 # A socket to which we forward summarized amavisd data.
 # should match a socket of the same name in amavis-status
