--- altermime-0.3-20100505/qpe.c.orig	2014-04-02 10:42:04.759363198 -0700
+++ altermime-0.3-20100505/qpe.c	2014-04-02 10:42:11.663362967 -0700
@@ -91,7 +91,7 @@
 	char CRLF[]="\r\n";
 
 
-	fprintf(stderr,"%s:%d:DEBUG: Line terminator length = %d", FL, strlen(line_terminator));
+	fprintf(stderr,"%s:%d:DEBUG: Line terminator length = %d", FL, (int)strlen(line_terminator));
 
 	if (line_terminator == NULL) line_terminator = CRLF;
 
