--- hotspot/src/os/linux/vm/os_linux.inline.hpp.orig	2018-09-25 14:42:40.744843438 +0200
+++ hotspot/src/os/linux/vm/os_linux.inline.hpp	2018-09-24 20:07:20.042699497 +0200
@@ -116,6 +116,11 @@
 
 inline struct dirent* os::readdir(DIR* dirp, dirent *dbuf)
 {
+// readdir_r has been deprecated since glibc 2.24.
+// See https://sourceware.org/bugzilla/show_bug.cgi?id=19056 for more details.
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wdeprecated-declarations"
+
   dirent* p;
   int status;
   assert(dirp != NULL, "just checking");
@@ -129,6 +134,8 @@
     return NULL;
   } else
     return p;
+
+#pragma GCC diagnostic pop
 }
 
 inline int os::closedir(DIR *dirp) {
