--- nginx/src/mail/ngx_mail_auth_http_module.c	2023-03-07 16:26:44.704430400 +0530
+++ nginx/src/mail/ngx_mail_auth_http_module.c	2023-09-14 18:47:12.605101100 +0530
@@ -1536,6 +1536,7 @@
         conf->host_header = prev->host_header;
         conf->uri = prev->uri;
 
+        /* zimbra uses zmauth, so no auth_http
         if (conf->peer == NULL) {
             ngx_log_error(NGX_LOG_EMERG, cf->log, 0,
                           "no \"auth_http\" is defined for server in %s:%ui",
@@ -1543,6 +1544,7 @@
 
             return NGX_CONF_ERROR;
         }
+        */
     }
 
     ngx_conf_merge_msec_value(conf->timeout, prev->timeout, 60000);
