--- nginx/src/mail/ngx_mail_imap_module.h	2023-03-07 16:26:44.722000700 +0530
+++ nginx/src/mail/ngx_mail_imap_module.h	2023-03-04 13:25:59.938112400 +0530
@@ -24,6 +24,15 @@
     ngx_uint_t   auth_methods;
 
     ngx_array_t  capabilities;
+
+    ngx_array_t  id_params;
+    ngx_str_t    id;
+
+    ngx_str_t    ua_name;    /* user agent name coming from ID field "NAME" */
+    ngx_str_t    ua_version; /* user agent version coming from ID field "VERSION" */
+
+    ngx_flag_t   literalauth;
+    ngx_str_t    greeting;
 } ngx_mail_imap_srv_conf_t;
 
 
