--- nginx/src/mail/ngx_mail_smtp_module.c	2023-03-07 16:26:44.739565600 +0530
+++ nginx/src/mail/ngx_mail_smtp_module.c	2023-09-14 18:47:12.670786200 +0530
@@ -50,7 +50,10 @@
 
     ngx_string("451 4.3.2 Internal server error" CRLF),
     ngx_string("421 4.7.1 SSL certificate error" CRLF),
-    ngx_string("421 4.7.1 No required SSL certificate" CRLF)
+    // Zimbra customizations start here (Jira Tickets: )
+    ngx_string("421 4.7.1 No required SSL certificate" CRLF),
+    ngx_string("") /* zimbra add a quite_msg field in mail protocol */
+    // Zimbra customizations end here
 };
 
 
