diff -ruN a/include/openssl/opensslv.h b/include/openssl/opensslv.h
--- a/include/openssl/opensslv.h	2021-04-14 03:14:12.720095390 +0000
+++ b/include/openssl/opensslv.h	2021-04-14 03:15:06.540096909 +0000
@@ -40,7 +40,7 @@
  *  major minor fix final patch/beta)
  */
 # define OPENSSL_VERSION_NUMBER  0x101010bfL
-# define OPENSSL_VERSION_TEXT    "OpenSSL 1.1.1k  25 Mar 2021"
+# define OPENSSL_VERSION_TEXT    "OpenSSL 1.1.1k FIPS 25 Mar 2021"
 
 /*-
  * The macros below are to be used for shared library (.so, .dll, ...)
