diff -ruN source/t/cz-03zlib-v1.t modified/t/cz-03zlib-v1.t
--- source/t/cz-03zlib-v1.t	2025-08-26 05:19:32.167074617 +0000
+++ modified/t/cz-03zlib-v1.t	2025-08-26 05:32:00.576747366 +0000
@@ -14,6 +14,8 @@
 use CompTestUtils;
 use Symbol;
 
+use constant ZLIB_1_2_12_0 => 0x12C0;
+
 BEGIN 
 { 
     # use Test::NoWarnings, if available
@@ -700,7 +702,16 @@
      
     ($GOT, $status) = $k->inflate($rest) ;
      
-    ok $status == Z_DATA_ERROR ;
+    # Z_STREAM_END returned by 1.12.2, Z_DATA_ERROR for older zlib
+    # always Z_STREAM_ENDin zlib_ng
+    if (ZLIB_VERNUM >= ZLIB_1_2_12_0 || Compress::Raw::Zlib::is_zlibng())
+    {
+        cmp_ok $status, '==', Z_STREAM_END ;
+    }
+    else
+    {
+        cmp_ok $status, '==', Z_DATA_ERROR ;
+    }
     ok $Z . $GOT eq $goodbye ;
 }
 
